x/staking/src/keeper/validator.rs

```diff
@@ -8,93 +15,248 @@ impl<
+                r#type: EVENT_TYPE_CREATE_VALIDATOR.to_string(),
+                attributes: vec![
+                    EventAttribute {
+                        key: ATTRIBUTE_KEY_VALIDATOR.as_bytes().into(),
+                        value: msg.validator_address.to_string().as_bytes().to_vec().into(),
+                        index: false,
+                    },
+                    EventAttribute {
+                        key: ATTRIBUTE_KEY_AMOUNT.as_bytes().into(),
+                        value: serde_json::to_vec(&msg.value)
+                            .expect(SERDE_ENCODING_DOMAIN_TYPE)
+                            .into(),
+                        index: false,
+                    },
+                ],
+            },
+            Event {
+                r#type: EVENT_TYPE_MESSAGE.to_string(),
+                attributes: vec![
+                    EventAttribute {
+                        key: ATTRIBUTE_KEY_MODULE.as_bytes().into(),
+                        value: ATTRIBUTE_VALUE_CATEGORY.as_bytes().to_vec().into(),
+                        index: false,
+                    },
+                    EventAttribute {
+                        key: ATTRIBUTE_KEY_SENDER.as_bytes().into(),
+                        value: msg.delegator_address.to_string().as_bytes().to_vec().into(),
+                        index: false,
+                    },
+                ],
+            },
+        ]);
+
+        Ok(())
+    }
+
+    pub fn validator<DB: Database, CTX: QueryableContext<DB, SK>>(
         &self,
         ctx: &CTX,
-        key: &[u8],
-    ) -> anyhow::Result<Validator> {
+        key: &ValAddress,
+    ) -> Option<Validator> {
         let store = ctx.kv_store(&self.store_key);
         let validators_store = store.prefix_store(VALIDATORS_KEY);
-        if let Some(e) = validators_store.get(key) {
-            Ok(serde_json::from_slice(&e)?)
+        validators_store
+            .get(key.to_string().as_bytes())
+            .map(|e| serde_json::from_slice(&e).expect(SERDE_DECODING_DOMAIN_TYPE))
+    }
+
+    pub fn set_validator<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        validator: &Validator,
+    ) {
+        let store = ctx.kv_store_mut(&self.store_key);
+        let mut validators_store = store.prefix_store_mut(VALIDATORS_KEY);
+        validators_store.set(
+            validator.operator_address.to_string().as_bytes().to_vec(),
+            serde_json::to_vec(&validator).expect(SERDE_ENCODING_DOMAIN_TYPE),
+        );
+    }
+
+    pub fn validator_by_cons_addr<DB: Database, CTX: QueryableContext<DB, SK>>(
+        &self,
+        ctx: &CTX,
+        addr: &ConsAddress,
+    ) -> anyhow::Result<Validator> {
+        let store = ctx.kv_store(&self.store_key);
+        let validators_store = store.prefix_store(VALIDATORS_BY_CONS_ADDR_KEY);
+
+        if let Some(bytes) = validators_store.get(addr.to_string().as_bytes()) {
+            Ok(serde_json::from_slice(&bytes)?)
```

Comment:
unwrap_or_corrupt?

Reply:
fixed

Reply:
I think that you misunderstood. Instead of matching, you should call `unwrap_or_corrupt` from `database` crate.

Reply:
> I think that you misunderstood. Instead of matching, you should call `unwrap_or_corrupt` from `database` crate.

I resolved the issue correctly. I used `map` method and behavior like `unwrap_or_corrupt` [here](https://github.com/rumos-io/gears/pull/97/commits/c0b6d417ddc3caf363990e49012a40e2ea1f3894#diff-4568d22f0e45db6b3ebb2b85321909388d0d164dfc3b71cc91fc42eff5a412cdR172). 

I've proposed to make implementation of trait for `Resutl<T, E>` above.